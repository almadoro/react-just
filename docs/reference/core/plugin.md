# Plugin (`react-just/vite`)

This Vite plugin enables React with React Server Components (RSC) support and provides a development server.

```ts [vite.config.ts] {1,5}
import react from "react-just/vite";
import { defineConfig } from "vite";

export default defineConfig({
  plugins: [react()],
});
```

## Options

```ts
react(options?: ReactJustOptions);

interface ReactJustOptions = {
  app?: string;
};
```

The plugin accepts an _optional_ `options` object with the following properties:

- `app`: Path to the module that exports the App Component. By default, first matching file among:
  - `src/index.tsx`
  - `src/index.jsx`
  - `src/index.ts`
  - `src/index.js`
