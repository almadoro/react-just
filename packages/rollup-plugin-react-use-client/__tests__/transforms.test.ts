import fs from "node:fs/promises";
import path from "node:path";
import { describe, expect, test } from "vitest";
import { transform } from "./plugin";

describe("transforms", () => {
  describe("export default declarations", () => {
    test(
      "class declarations",
      transformTest("export-default-declaration/class-declaration.js"),
    );

    test(
      "function declarations",
      transformTest("export-default-declaration/function-declaration.js"),
    );

    test(
      "identifier",
      transformTest("export-default-declaration/identifier.js"),
    );

    test(
      "value declarations",
      transformTest("export-default-declaration/value-declaration.js"),
    );
  });

  describe("export named declarations", () => {
    test(
      "array pattern",
      transformTest("export-named-declaration/array-pattern.js"),
    );

    test(
      "assignment pattern",
      transformTest("export-named-declaration/assignment-pattern.js"),
    );

    test("class", transformTest("export-named-declaration/class.js"));

    test("identifier", transformTest("export-named-declaration/identifier.js"));

    test("function", transformTest("export-named-declaration/function.js"));

    test(
      "object pattern",
      transformTest("export-named-declaration/object-pattern.js"),
    );

    test(
      "rest element",
      transformTest("export-named-declaration/rest-element.js"),
    );
  });

  test(
    "export named from source",
    transformTest("export-named-from-source.js"),
  );

  test("export named specifiers", transformTest("export-named-specifiers.js"));

  test("export all declarations not supported", () => {
    expect(
      transform("'use client'; export * from 'pkg';", "module.js"),
    ).rejects.toThrowError();
  });
});

function transformTest(filePath: string) {
  return async () => {
    const input = await fs.readFile(
      path.resolve(import.meta.dirname, "fixtures/transforms", filePath),
      "utf-8",
    );

    const output = await transform(input, "module.js");

    expect(output?.code).toMatchSnapshot();
  };
}
