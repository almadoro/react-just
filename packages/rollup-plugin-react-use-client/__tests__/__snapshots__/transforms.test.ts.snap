// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transforms > export default declarations > class declarations 1`] = `
"import {registerClientReference} from "react";
class A {}
const $$Ref$$default = registerClientReference(A, "module.js", "default");
export default $$Ref$$default;
"
`;

exports[`transforms > export default declarations > function declarations 1`] = `
"import {registerClientReference} from "react";
function $$default$$() {}
const $$Ref$$default = registerClientReference($$default$$, "module.js", "default");
export default $$Ref$$default;
"
`;

exports[`transforms > export default declarations > identifier 1`] = `
"import {registerClientReference} from "react";
let a;
const $$Ref$$default = registerClientReference(a, "module.js", "default");
export default $$Ref$$default;
"
`;

exports[`transforms > export default declarations > value declarations 1`] = `
"import {registerClientReference} from "react";
const $$default$$ = 1;
const $$Ref$$default = registerClientReference($$default$$, "module.js", "default");
export default $$Ref$$default;
"
`;

exports[`transforms > export named declarations > array pattern 1`] = `
"import {registerClientReference} from "react";
let [a, {b}, [c], d = 1, , ...e] = [];
const $$Ref$$a = registerClientReference(a, "module.js", "a");
const $$Ref$$b = registerClientReference(b, "module.js", "b");
const $$Ref$$c = registerClientReference(c, "module.js", "c");
const $$Ref$$d = registerClientReference(d, "module.js", "d");
const $$Ref$$e = registerClientReference(e, "module.js", "e");
export {$$Ref$$a as a, $$Ref$$b as b, $$Ref$$c as c, $$Ref$$d as d, $$Ref$$e as e};
"
`;

exports[`transforms > export named declarations > assignment pattern 1`] = `
"import {registerClientReference} from "react";
let [a = 1, {b} = c, [d] = e] = [];
const $$Ref$$a = registerClientReference(a, "module.js", "a");
const $$Ref$$b = registerClientReference(b, "module.js", "b");
const $$Ref$$d = registerClientReference(d, "module.js", "d");
export {$$Ref$$a as a, $$Ref$$b as b, $$Ref$$d as d};
"
`;

exports[`transforms > export named declarations > class 1`] = `
"import {registerClientReference} from "react";
class A {}
const $$Ref$$A = registerClientReference(A, "module.js", "A");
export {$$Ref$$A as A};
"
`;

exports[`transforms > export named declarations > function 1`] = `
"import {registerClientReference} from "react";
function a() {}
const $$Ref$$a = registerClientReference(a, "module.js", "a");
export {$$Ref$$a as a};
"
`;

exports[`transforms > export named declarations > identifier 1`] = `
"import {registerClientReference} from "react";
let a;
const $$Ref$$a = registerClientReference(a, "module.js", "a");
export {$$Ref$$a as a};
"
`;

exports[`transforms > export named declarations > object pattern 1`] = `
"import {registerClientReference} from "react";
let {a, b: {c}, d: [e], f = 1, ...g} = {};
const $$Ref$$a = registerClientReference(a, "module.js", "a");
const $$Ref$$c = registerClientReference(c, "module.js", "c");
const $$Ref$$e = registerClientReference(e, "module.js", "e");
const $$Ref$$f = registerClientReference(f, "module.js", "f");
const $$Ref$$g = registerClientReference(g, "module.js", "g");
export {$$Ref$$a as a, $$Ref$$c as c, $$Ref$$e as e, $$Ref$$f as f, $$Ref$$g as g};
"
`;

exports[`transforms > export named declarations > rest element 1`] = `
"import {registerClientReference} from "react";
let [...a] = [];
let [...[b, c]] = [];
let [...{d: e}] = [];
const $$Ref$$a = registerClientReference(a, "module.js", "a");
export {$$Ref$$a as a};
const $$Ref$$b = registerClientReference(b, "module.js", "b");
const $$Ref$$c = registerClientReference(c, "module.js", "c");
export {$$Ref$$b as b, $$Ref$$c as c};
const $$Ref$$e = registerClientReference(e, "module.js", "e");
export {$$Ref$$e as e};
"
`;

exports[`transforms > export named from source 1`] = `
"import {registerClientReference} from "react";
import {a as $$a$$, b as $$c$$, b as $$d$$, default as $$default$$, default as $$e$$} from "pkg";
let c = 1;
const $$Ref$$a = registerClientReference($$a$$, "module.js", "a");
const $$Ref$$c = registerClientReference($$c$$, "module.js", "c");
const $$Ref$$d = registerClientReference($$d$$, "module.js", "d");
const $$Ref$$default = registerClientReference($$default$$, "module.js", "default");
const $$Ref$$e = registerClientReference($$e$$, "module.js", "e");
export {$$Ref$$a as a, $$Ref$$c as c, $$Ref$$d as d, $$Ref$$default as default, $$Ref$$e as e};
"
`;

exports[`transforms > export named specifiers 1`] = `
"import {registerClientReference} from "react";
let a;
let b;
let d;
const $$Ref$$a = registerClientReference(a, "module.js", "a");
const $$Ref$$c = registerClientReference(b, "module.js", "c");
const $$Ref$$default = registerClientReference(d, "module.js", "default");
export {$$Ref$$a as a, $$Ref$$c as c, $$Ref$$default as default};
"
`;

exports[`transforms export default declarations class declarations 1`] = `
"import {registerClientReference} from "react";
class __Impl__default__A {}
export default registerClientReference(__Impl__default__A, "module.js", "default");
"
`;

exports[`transforms export default declarations function declarations 1`] = `
"import {registerClientReference} from "react";
function __Impl__default__() {}
export default registerClientReference(__Impl__default__, "module.js", "default");
"
`;

exports[`transforms export default declarations identifier 1`] = `
"import {registerClientReference} from "react";
let __Impl__a;
export default registerClientReference(__Impl__a, "module.js", "default");
"
`;

exports[`transforms export default declarations value declarations 1`] = `
"import {registerClientReference} from "react";
const __Impl__default = 1;
export default registerClientReference(__Impl__default, "module.js", "default");
"
`;
